akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala

```diff
@@ -3,70 +3,464 @@
+
+    val in = Inlet[Any]("in")
 
-    override protected def newInstance(shape: SinkShape[Nothing]): SinkModule[Nothing, Future[Attributes]] =
-      new AttributesSink(attributes, shape)
+    override protected def initialAttributes: Attributes = _initialAttributes
+    override val shape = SinkShape(in)
+    override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Attributes) = {
+      val logic = new GraphStageLogic(shape) {
+        override def preStart(): Unit = {
+          pull(in)
+        }
+        setHandler(in, new InHandler {
+          override def onPush(): Unit = {
+            grab(in)
+            pull(in)
+          }
+        })
+      }
 
-    override def withAttributes(attr: Attributes): SinkModule[Nothing, Future[Attributes]] =
-      new AttributesSink(attr, amendShape(attr))
+      (logic, inheritedAttributes)
+    }
   }
 
+  class ThreadNameSnitchingStage(initialDispatcher: String) extends GraphStage[SourceShape[String]] {
+    val out = Outlet[String]("out")
+    override val shape = SourceShape.of(out)
+    override protected def initialAttributes: Attributes = ActorAttributes.dispatcher(initialDispatcher)
+    def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+      setHandler(out, new OutHandler {
+        def onPull(): Unit = {
+          push(out, Thread.currentThread.getName)
+          completeStage()
+        }
+      })
+    }
+
+  }
+
+  def whateverAttribute(label: String): Attributes = Attributes(WhateverAttribute(label))
+  case class WhateverAttribute(label: String) extends Attribute
 }
 
-class AttributesSpec extends StreamSpec {
+class AttributesSpec extends StreamSpec(ConfigFactory.parseString(
+  """
+    my-dispatcher {
+      type = Dispatcher
+      executor = "thread-pool-executor"
+      thread-pool-executor {
+        fixed-pool-size = 1
+      }
+      throughput = 1
+    }
+  """).withFallback(Utils.UnboundedMailboxConfig)) {
+
   import AttributesSpec._
 
   val settings = ActorMaterializerSettings(system)
     .withInputBuffer(initialSize = 2, maxSize = 16)
 
   implicit val materializer = ActorMaterializer(settings)
 
-  "attributes" must {
+  "an attributes instance" must {
+
+    val attributes = Attributes.name("a") and Attributes.name("b") and Attributes.inputBuffer(1, 2)
+
+    "give access to the least specific attribute" in {
+      attributes.leastSpecific[Name] should ===(Some(Attributes.Name("a")))
+    }
+
+    "give access to the most specific attribute value" in {
+      attributes.mostSpecific[Name] should ===(Some(Attributes.Name("b")))
+    }
+
+    "prefer most specific important attribute to most specific attribute" in {
+      val attributes = Attributes.name("a")
+        .and(Attributes.name("b").important)
+        .and(Attributes.name("c"))
+        .and(Attributes.name("d").important)
```

Comment:
Ah, I see this replaced the thing or what happened here?

Reply:
last added attribute on the same graph is most specific

Reply:
meh that looks confusing, but I guess noone really usees this api like this directly hmm

Reply:
This is just a shortcut to do what the graph stage does when you add them on different levels (or the same but that will probably not be used like this very often except for when adding an attribute on a graph and returning it from a function, then you still want this behavior if someone else adds the same attribute after)